### PR TITLE
Implements complex isfinite and isinf

### PR DESCRIPTION
**Pull request by mruberry**
Implements complex isfinite and isinf, consistent with NumPy.

A complex value is finite if and only if both its real and imaginary part are finite. 

A complex value is infinite if and only if its real or imaginary part are infinite. 

Old isfinite, isinf, and isnan tests are modernized and instead of fixtures the torch results are compared with NumPy. A new test is added for complex isfinite, isinf, and isnan. The docs for each function are updated to clarify what finite, infinite, and NaN values are.

The new tests rely on a new helper, _np_compare, that we'll likely want to generalize in the near future and use in more tests. 

Addresses part of the complex support tasks. See https://github.com/pytorch/pytorch/issues/33152.

### CODE REVIEW

File: torch/_torch_docs.py [hunk 1]
```diff
@@ -2580,6 +2580,7 @@ def merge_dicts(*dicts):
 add_docstr(torch.isinf,
            r"""
 Returns a new tensor with boolean elements representing if each element is `+/-INF` or not.
+Complex values are infinite when their real or imaginary part is infinite.
```
Comment (anjali411) [inline torch/_torch_docs.py:2583]: nit -  `real and(or) imaginary part is infinite`

File: torch/_torch_docs.py [hunk 2]
```diff
@@ -2612,6 +2616,7 @@ def merge_dicts(*dicts):
 add_docstr(torch.isnan,
            r"""
 Returns a new tensor with boolean elements representing if each element is `NaN` or not.
+Complex values are considered `NaN` when either their real or imaginary part is NaN.
```
Comment (anjali411) [inline torch/_torch_docs.py:2619]: nit - `when their real and(or) imaginary part is NaN.`

**Review by anjali411 — APPROVED** [review]
LGTM :) thanks!

**Review by ngimel — APPROVED** [review]

File: aten/src/ATen/native/TensorCompare.cpp [hunk 3]
```diff
@@ -62,20 +62,36 @@ Tensor isnan(const Tensor& self) {
...
-  // Integral tensor types are finite
-  if (!self.is_floating_point()) {
+  // Note: Integral tensor values are always finite
+  if (c10::isIntegralType(self.scalar_type(), /*include_bool=*/true)) {
     return at::ones_like(self, at::kBool, at::MemoryFormat::Preserve);
   }
+
+  // Note: a complex value is finite iff both parts are finite
+  if (self.is_complex()) {
+    const auto float_type = c10::toValueType(self.scalar_type());
```
Comment (ngimel) [inline aten/src/ATen/native/TensorCompare.cpp:90]: `self.abs().to(float_type).isfinite()` is faster and achieves the same result.

File: torch/_torch_docs.py (hunk 1 above)
Comment (mruberry) [reply torch/_torch_docs.py:2583]: Sure.

File: aten/src/ATen/native/TensorCompare.cpp [hunk 4]
```diff
@@ -62,20 +62,35 @@ Tensor isnan(const Tensor& self) {
...
+    return at::isfinite(self.abs().to(float_type));
```
Comment (anjali411) [inline aten/src/ATen/native/TensorCompare.cpp:91]: ahh we should have updated this to just .abs()

**Pull request closed without merging**